### ISSUE TITLE
Output streams not properly synchronized on Python 3 script

### ISSUE THREAD

**Issue by DangerOnTheRanger** [xonsh/xonsh#2561]
How to reproduce on xonsh 0.6:
- Download https://gist.github.com/DangerOnTheRanger/656a99274c677aac5c4ca12855670278
- Run it under xonsh
- Hit Enter, or just give any valid/invalid input
- You'll get some output like the following:

https://i.imgur.com/MMVKmic.png

I should note that the error seems to be consistent, but non-deterministic; I can't get the same incorrect output twice. I'm not familiar with xonsh's internals enough to figure out why this would be the case, but it certainly couldn't hurt.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by gforsyth** [comment]
Thanks for reporting @DangerOnTheRanger! 

As @scopatz noted on gitter, this is almost certainly an issue with how stderr and stdout are being synchronized.

**Comment by DangerOnTheRanger** [comment]
Chiming back in to report that this is still an issue with 0.8.1 - I'm just now reading @scopatz's https://github.com/xonsh/xonsh/issues/2129#issuecomment-283261329; I'd be willing to implement ordering if someone else hasn't already done so.

**Comment by scopatz** [comment]
@DangerOnTheRanger - Please! We could defintiely use the help here!  Let us know if you have any questions

**Comment by daringer** [comment]
Hey all, just wanted to confirm that I still can see the very same issue:

Versions:
* xonsh: latest git 
* (x)console: kitty 0.14.3
* distribution: Arch Linux

Overall this is quite annoying (at least until one knows about this), for me this happens especially for `pacman` confirmations. I was wondering whether this is an issues that happens more often for some people than others, because I personally encounter it a lot.

@DangerOnTheRanger can you already share some progress? Would be willing to support/test, whatever needed...

**Comment by astronouth7303** [comment]
This is a limitation of posix: You cannot synchronize two different Pipes (streams). We could write a little C fastloop to try to read them both as fast as possible, but it'd still be possible for them to desync.

The way most shells do it is that a commands `stdout` and `stderr` actually point to the same pipe, so they're both writing into the same kernel buffer.

**Comment by daringer** [comment]
For a command started within the context of a shell this redirection into a "common pipe" feels valid, as an user is always expecting exactly this behavior?

Further a shell command reading from any of these both pipes has to be indicated using some "keyword" (character combination) which might then internally lead to not only writing the stderr data into the "common pipe" but additionally (or instead) to the other receiver. 

Or am I thinking to naive here?

Apart from the technical challenge here, my impression is that this is a bug which surely scares away people from xonsh, because if one doesn't know what's behind it, it feels like a massive flaw during daily shell usage.

**Comment by astronouth7303** [comment]
It is.

The big reason xonsh does this is for storing stdout in history.

**Comment by daringer** [comment]
Because this is really annoying me, I tried the most primitive solution first.

using this commit: https://github.com/daringer/xonsh/commit/960eebb19f6e97f3d9fd178e1854a03a25fccd1c
(there might be unforeseeable side effects due to the hacky-ness of this fix)

One might set the env variable `MERGE_STDERR` to something evaluating to `True` and `stderr` will be written into `stdout`. Very likely this will have evil side-effects, but despite this, at least there is some way to explicitly enforce this behavior.

Example output before and after setting the env var: https://pastebin.com/TA1H52sN

Obviously this is not the "correct" solution overall, still delivers, at least on the first glance. Once pipes in chained commands come into play the outcome won't be nice for the receiver side expecting stdout but also getting stderr mixed in. But on the other side it is known to the `SubprocSpec::run()` method that a pipe is waiting for its stdout, right? Thus just maintain a second copy of `stdout` to use just this for a requested pipe? 

Any ideas which other problems would we get with such a "dirty" solution for this issue?

cheers

**Comment by scopatz** [comment]
This is kind of a clever solution.  Obviously, it shouldn't be the default, but I am not oppossed to something like this.  A better implemention of this would instead set a default redirection when no redirects are specified. Then you could have a default of "err>out". It would also make something like "out>out" meaningful.

**Comment by scopatz** [comment]
This could also be used in script to always redirect command output to a file.

**Comment by scopatz** [comment]
I would love to see a PR for this

**Comment by daringer** [comment]
@scopatz, nice to hear, for sure I'll like to provide a PR, currently a little swamped, but can see the light already. Despite that, I (for now) fully close both eyes while reading @astronouth7303's comment :sweat_smile:. The shown solution is clearly not the *right* one (no idea actually, but likely due to my lack of understanding what and why `xonsh` is different from what `bash`, `tcsh`, `zsh`, `ipython` are doing and if they redirect all `stderr` to `stdout`, do they?), but apart from the missing fancyness of this solution. It already improved daily console life, at least for me. 

@scopatz do I get you right, you would suggest to have something like a generic redirection in-between any of `std-in/out/err`, which might then be changed by the user explicitly, right? 
Script use-case understood, but your "out>out" example confuses me, do you imply redirecting the `stdout` from one application directly to another one (connected via a pipe?), thus both are writing to `stdout` in parallel?

**Comment by scopatz** [comment]
`out>out` would just be a signifier that stdout should go to the usual place and not be redirected.

**Comment by jnoortheen** [comment]
I have checked the gist script in xonsh latest version. It is solved when default behaviour of capturing output from subprocess is turned off. https://github.com/xonsh/xonsh/issues/2129
